5calls/app/src/main/java/org/a5calls/android/a5calls/controller/LocationActivity.java

```diff
@@ -228,21 +219,53 @@ public void onProviderDisabled(String provider) {
                 }
             };
             locationManager.requestLocationUpdates(provider, 10, 0, mLocationListener);
+
         } else {
-            updateSharedPrefs(location);
+            onReceiveLocation(location);
         }
     }
 
-    private void updateSharedPrefs(Location location) {
-        accountManager.setLat(this, String.valueOf(location.getLatitude()));
-        accountManager.setLng(this, String.valueOf(location.getLongitude()));
+    private void onSubmitZip(String zip) {
+        // Is it a string that is exactly 5 characters long?
+        if (TextUtils.isEmpty(zip) || zip.length() != 5) {
+            zipEdit.setError(getResources().getString(R.string.zip_error));
+            return;
+        }
+        try {
+            // Make sure it is a number, too, by trying to parse it.
+            Integer.parseInt(zip);
+        } catch (NumberFormatException e) {
+            zipEdit.setError(getResources().getString(R.string.zip_error));
+            return;
+        }
+        // If we made it here, the zip is valid! Update the UI and send the request.
+        accountManager.setZip(this, zip);
+
+        // Delete latlng, because the user specifically requested a zip and we default to lat/long.
+        accountManager.setLat(this, null);
+        accountManager.setLng(this, null);
+
         returnToMain();
     }
 
-    private void setFromMain() {
-        // If we already have a zip, this isn't our first time to the app, so we can go "up"
-        // to return to the main activity.
-        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
-        mFromMain = true;
+
+    private void onReceiveLocation(Location location) {
+        accountManager.setLat(this, String.valueOf(location.getLatitude()));
+        accountManager.setLng(this, String.valueOf(location.getLongitude()));
+
+        // Update the zip field from the location
+        if (location != null) {
+            Geocoder geocoder = new Geocoder(this, Locale.getDefault());
+            try {
+                List<Address> addresses = geocoder.getFromLocation(location.getLatitude(), location.getLongitude(), 1);
+                if (addresses.size() > 0) {
+                    accountManager.setZip(this, addresses.get(0).getPostalCode());
```

Comment:
I don't think the address can be null, but according to the docs, the list can be null.  I've added a check for that.